packages/mui-material/src/Button/Button.js

```diff
@@ -405,16 +607,32 @@ const Button = React.forwardRef(function Button(inProps, ref) {
       ownerState={ownerState}
       className={clsx(contextProps.className, classes.root, className, positionClassName)}
       component={component}
-      disabled={disabled}
+      disabled={disabled || loading}
       focusRipple={!disableFocusRipple}
       focusVisibleClassName={clsx(classes.focusVisible, focusVisibleClassName)}
       ref={ref}
       type={type}
+      id={id}
       {...other}
       classes={classes}
     >
       {startIcon}
-      {children}
+      {loading && ownerState.loadingPosition === 'end' ? (
+        <ButtonLoadingLabel className={classes.label} ownerState={ownerState}>
+          {children}
+        </ButtonLoadingLabel>
+      ) : (
+        buttonLoadingIndicator
+      )}
+
+      {loading && ownerState.loadingPosition === 'end' ? (
+        buttonLoadingIndicator
+      ) : (
+        <ButtonLoadingLabel className={classes.label} ownerState={ownerState}>
+          {children}
+        </ButtonLoadingLabel>
+      )}
```

Comment:
I think this can be considered a breaking change since we're wrapping `children` with a `<span>`.

Also, we can make the conditions easier to follow. I think this is equivalent:

```suggestion
      {loading && ownerState.loadingPosition === 'start' ? buttonLoadingIndicator : null}
      <ButtonLoadingLabel className={classes.label} ownerState={ownerState}>
          {children}
      </ButtonLoadingLabel>
      {loading && ownerState.loadingPosition === 'end' ? buttonLoadingIndicator : null}
```

and change above this so `buttonLoadingIndicator` always has a value:

```diff
-const buttonLoadingIndicator = loading ? (
+const buttonLoadingIndicator = (
   <ButtonLoadingIndicator className={classes.loadingIndicator} ownerState= {ownerState}>
     {loadingIndicator}
   </ButtonLoadingIndicator>
-) : null;
+);
```

Reply:
This won't work when `loadingPosition` is `center`. I could try the following:

```jsx
{loading && ownerState.loadingPosition === 'start' ? buttonLoadingIndicator : null}

{ownerState.loadingPosition !== 'center' ? (
  <ButtonLoadingLabel className={classes.label} ownerState={ownerState}>
    {children}
  </ButtonLoadingLabel>
) : (
  loading && buttonLoadingIndicator
)}

{loading && ownerState.loadingPosition === 'end' ? buttonLoadingIndicator : null}
```

But this breaks the button UI. Currently, when `loadingPosition` is `center`, both the loading indicator and label are added to the DOM, with the label hidden using `color: transparent` (See [styles](https://github.com/mui/material-ui/pull/42987/files#diff-2e6df3f03bdd7cef475f4362270feda335c131b3acb55ba73b0f33b3cd4ea357R328-R330)). I know this is not ideal but it preserves the layout without breaking. This approach matches the previous `LoadingButton` behavior, so it won’t introduce any changes. 

I've simplified it like this:

```jsx
{loading && (ownerState.loadingPosition === 'start' || ownerState.loadingPosition === 'center') ? buttonLoadingIndicator : null}
<ButtonLoadingLabel className={classes.label} ownerState={ownerState}>
  {children}
</ButtonLoadingLabel>
{loading && ownerState.loadingPosition === 'end' ? buttonLoadingIndicator : null}
```

----

> I think this can be considered a breaking change since we're wrapping children with a `span`.

We can wrap `children` in a `span` only when loading. When not loading, it won’t be wrapped, as before, so this shouldn’t be a breaking change. `span` was added in #35198.